### PR TITLE
chore: optimise remove windows, adding test

### PR DESCRIPTION
**Pull request by yhl25**
Signed-off-by: Yashash H L <yashashhl25@gmail.com>

### CODE REVIEW

File: pkg/window/strategy/sliding/sliding.go [hunk 1]
```diff
@@ -161,6 +161,8 @@ func (s *Sliding) RemoveWindows(wm time.Time) []window.AlignedKeyedWindower {
 		if win.EndTime().Before(wm) {
 			s.entries.Remove(e)
 			closedWindows = append(closedWindows, win)
+		} else {
```
Comment (vigith) [inline pkg/window/strategy/sliding/sliding.go:164]: This for loop will only run for 1 iteration, is that expected?

Comment (yhl25) [reply pkg/window/strategy/sliding/sliding.go:164]: this for loop will run until you find a window with end time > watermark

**Review by vigith — APPROVED** [review]

File: pkg/reduce/readloop/readloop.go [hunk 2]
```diff
@@ -94,6 +95,7 @@ func NewReadLoop(ctx context.Context,
 		whereToDecider:        whereToDecider,
 		publishWatermark:      pw,
 		udfInvocationTracking: make(map[partition.ID]*task),
+		latestWatermark:       processor.Watermark(time.Time{}),
```
Comment (ashwinidulams) [inline pkg/reduce/readloop/readloop.go:98]: rename this to indicate that you are tracking the window boundary

File: pkg/reduce/readloop/readloop.go [hunk 3]
```diff
@@ -150,8 +152,15 @@ func (rl *ReadLoop) Process(ctx context.Context, messages []*isb.ReadMessage) {
 	// for each successfully written message,
 	for _, m := range successfullyWrittenMessages {
 		// close any windows that need to be closed.
+		var closedWindows []window.AlignedKeyedWindower
 		wm := processor.Watermark(m.Watermark)
-		closedWindows := rl.windower.RemoveWindows(time.Time(wm))
+
+		// check to see if any windows can be closed, when the watermark is moved
+		if wm.After(time.Time(rl.latestWatermark)) {
```
Comment (vigith) [inline pkg/reduce/readloop/readloop.go:159]: Why are we comparing 2 watermarks? Isn't it obvious that those 2 will be diverging (the old one and the new one)?

File: pkg/reduce/readloop/readloop.go (hunk 2 above)
Comment (yhl25) [reply pkg/reduce/readloop/readloop.go:98]: I am not tracking the window boundary, I am tracking the latest watermark we have received

File: pkg/reduce/readloop/readloop.go (hunk 3 above)
Comment (yhl25) [reply pkg/reduce/readloop/readloop.go:159]: they can be same, I just wanted to invoke the remove windows function only when the watermark is progressed

**Comment by vigith** [comment]
Let's do a benchmark to compare the performance and allocation improvements.

**Comment by yhl25** [comment]
> Let's do a benchmark to compare the performance and allocation improvements.

I ran a benchmark test
observations:
I see better performance when RemoveWindows is invoked once per batch since we avoid a lot of comparisons with millisecond watermark progression
Not much difference if the watermark progresses slowly because we have a check inside remove windows which returns early.

File: pkg/reduce/readloop/readloop.go [hunk 4]
```diff
@@ -144,21 +144,26 @@ func (rl *ReadLoop) Process(ctx context.Context, messages []*isb.ReadMessage) {
 		rl.log.Errorw("Failed to write messages", zap.Int("totalMessages", len(messages)), zap.Int("writtenMessage", len(successfullyWrittenMessages)))
 	}
 
+	if len(successfullyWrittenMessages) == 0 {
```
Comment (vigith) [inline pkg/reduce/readloop/readloop.go:147]: why will it be == 0? We have checked early in the reduce code to continue only if there are `Read` messages. What we read should have been written, right?

Comment (yhl25) [reply pkg/reduce/readloop/readloop.go:147]: yeah, removed the check

**Review by vigith — APPROVED** [review]

**Pull request merged**